### ISSUE TITLE
Provide a way to parametrize aggregators and convergence criteria

### ISSUE THREAD

**Issue by vasia** [stratosphere/stratosphere#731]
Currently, there is no way to pass parameters to an aggregator or a convergence criterion. @StephanEwen has suggested to add support for this by changing the code for aggregators and convergence criterion to take an object, rather than a class, i.e. instead of 
`iteration.registerAggregator("my aggregator", Customaggregator.class)`
do something like this:
`iteration.registerAggregator("my aggregator", new CustomAggregator(params));`

I can start working on this, but I have a question: What should the type of `params` be?

**Comment by zentol** [comment]
I don't see how the type (and number) of params are relevant for this issue.

**Comment by StephanEwen** [comment]
I agree, Params can actually be arbitary. I would make it a regular java
class, and an extension of IOReadableWritable for the start, to deal with
serialization. The later, we can think about getting rid of the
IOReadbaleWritable.

**Comment by vasia** [comment]
Thanks, that's what I meant by type. I was actually thinking of Configuration, but I can just make an extension of IOReadableWritable, sure.

**Comment by StephanEwen** [comment]
Sorry, what I said was confusing. I would not make params an extension of
IOReadableWritable. The same way as MapFunction or JoinFunction can have a
constructors with arbitrary parameters, the Aggregator can have a
constructor with arbitrary parameters.

**Comment by vasia** [comment]
OK, I think I understand now. 
I assumed that since there are several common aggregators that have standard implementations, one would want a common way of instantiating them, in order to avoid defining multiple constructors or sub-classing.
For example, several programs will use the already provided LongSumAggregator, but if they need a different number of parameters, they would either have to add a constructor or subclass.
Do I make any sense? o.O

**Comment by zentol** [comment]
yes you do make sense. but consider this: if someone uses the LongSumAggregator, but wants to use additional parameters to do _something different_, he has to create a subclass anyway to add that new behaviour.

**Comment by vasia** [comment]
Hi,
while working on this issue, I noticed that there are no `getAggregators()`, `registerAggregator()` and `registerAggregationConvergenceCriterion()` methods in `DeltaIterativeDataSet`. Shall I implement these?

**Comment by StephanEwen** [comment]
Good idea, please do so!

**Comment by vasia** [comment]
Hey,
there exists a check in `NepheleJobGraphGenerator` that throws the following error if a custom convergence criterion is set in a DeltaIteration: "Cannot use custom convergence criterion with workset iteration. Workset iterations have implicit convergence criterion where workset is empty."
What is the logic behind this and should I remove it?

**Comment by StephanEwen** [comment]
The Delta Iteration terminates when the workset is empty. That is its
Termination Criterion and it comes with an Aggregator, which counts the
number of elements in the workset.

If you would set another TerminationCriterion, you would override that
"workset empty" termination criterion.

It makes sense to have more Aggregators available in the delta iterations.
In order to set an additional Termination Critertion, we would need to
change the code such that multiple criterions are allowed and that
convergence happens once one of them signals so.

For a start, I would suggest to allow the registration of additional
Aggregators, but not additional convergence criteria.

**Comment by vasia** [comment]
OK, so should I just implement the `registerAggregator()` method in the DeltaIterativeDataSet or also the `getAggregators()`? The 2nd one will also allow un-registering aggregators and get all registered aggregators, but also has a `registerAggregationConvergenceCriterion()` method (which will give an exception). 
I will open an issue for additional termination criteria in delta iteration, for future consideration :)

**Comment by StephanEwen** [comment]
Sounds good!
